### PR TITLE
Spreadsheet: Conditional Formatting

### PR DESCRIPTION
**Pull request by alimpfard**
Images, as I have no idea how to describe this:
shot-2020-09-23_23-50-33
shot-2020-09-23_23-51-33

---

I cannot get `ScrollableWidget` to play nice and lay these conditional formatting thingies out vertically, some help on that would be nice.
Also the GUI looks rather crap, hopefully I can put my GUI making hat on again and make a styling dialog/preview...eventually.

### CODE REVIEW

**Comment by asynts** [comment]
I didn't look through the code but spotted this:

Shouldn't the foreground color be white? At least that is what you selected in the other image.

**Comment by alimpfard** [comment]
> Shouldn't the foreground color be white

nothing is selected, it just shows white since there's no "nothing selected" colour.
that's partially what I'm referencing when I talk about a decent styling preview...

**Comment by alimpfard** [comment]
I made the colour preview widget fill the remaining space, to at least get an indicator of "no colour set" at the start, with the preview editor's text being empty...

File: Applications/Spreadsheet/CellType/Type.h [hunk 1]
```diff
@@ -52,8 +52,13 @@ class CellType {
     virtual JS::Value js_value(Cell&, const CellTypeMetadata&) const = 0;
     virtual ~CellType() { }
 
+    const StringView& name() const { return m_name; }
+
 protected:
     CellType(const StringView& name);
+
+private:
+    StringView m_name;
```
Comment (awesomekling) [inline Applications/Spreadsheet/CellType/Type.h:61]: This feels sketchy. Who guarantees that the data inside this `StringView` doesn't disappear?

Comment (alimpfard) [reply Applications/Spreadsheet/CellType/Type.h:61]: The fact that all the cell types are meant to be constructed once, and are static (in `CellType/Type.cpp`)?
I _could_ make it allocate a String, but I don't think anyone would even _want_ to instantiate one of these

File: Applications/Spreadsheet/CellTypeDialog.cpp [hunk 2]
```diff
@@ -403,4 +425,90 @@ CellTypeMetadata CellTypeDialog::metadata() const
     return metadata;
 }
 
+ConditionView::ConditionView(ConditionalFormat& fmt)
+    : m_format(fmt)
+{
+    load_from_json(cond_fmt_view_ui_json);
+    static_cast<GUI::Label*>(find_descendant_by_name("if_label"))->set_text("if...");
+    static_cast<GUI::Label*>(find_descendant_by_name("fg_color_label"))->set_text("Foreground Color...");
+    static_cast<GUI::Label*>(find_descendant_by_name("bg_color_label"))->set_text("Background Color...");
```
Comment (awesomekling) [inline Applications/Spreadsheet/CellTypeDialog.cpp:434]: You can set these in the JSON :^)

File: Applications/Spreadsheet/CellTypeDialog.cpp [hunk 3]
```diff
@@ -403,4 +425,90 @@ CellTypeMetadata CellTypeDialog::metadata() const
...
+
+    auto& fg_input = *static_cast<GUI::ColorInput*>(find_descendant_by_name("foreground_input"));
+    auto& bg_input = *static_cast<GUI::ColorInput*>(find_descendant_by_name("background_input"));
+    auto& formula_editor = *static_cast<GUI::TextEditor*>(find_descendant_by_name("formula_editor"));
+
+    formula_editor.set_tooltip("Use 'value' to refer to the current cell's value");
```
Comment (awesomekling) [inline Applications/Spreadsheet/CellTypeDialog.cpp:440]: You can set this in the JSON as well.

File: Applications/Spreadsheet/CellTypeDialog.cpp (hunk 2 above)
Comment (alimpfard) [reply Applications/Spreadsheet/CellTypeDialog.cpp:434]: Unfortunately, `text` does not apply to `GUI::Label`, and registering it made the linker unhappy.
I kinda forgot about that one actually ^^'

Comment (awesomekling) [reply Applications/Spreadsheet/CellTypeDialog.cpp:434]: a5878175e25bf28265d1358c2fdb1cc4b66cc61e ^_^

Comment (alimpfard) [reply Applications/Spreadsheet/CellTypeDialog.cpp:434]: oh.
I'll change it lol

File: Applications/Spreadsheet/CellType/Type.h (hunk 1 above)
Comment (alimpfard) [reply Applications/Spreadsheet/CellType/Type.h]: Less sketch for a few more allocs :shrug:

File: Applications/Spreadsheet/CellType/Type.h [hunk 4]
```diff
@@ -52,13 +52,13 @@ class CellType {
     virtual JS::Value js_value(Cell&, const CellTypeMetadata&) const = 0;
     virtual ~CellType() { }
 
-    const StringView& name() const { return m_name; }
+    const String& name() const { return m_name; }
```
Comment (alimpfard) [inline Applications/Spreadsheet/CellType/Type.h:55]: ugh why is this here
